src/hotspot/cpu/riscv/nativeInst_riscv.hpp

```diff
@@ -103,98 +102,46 @@ class NativeInstruction {
   }
 };
 
-inline NativeInstruction* nativeInstruction_at(address addr) {
+NativeInstruction* nativeInstruction_at(address addr) {
   return (NativeInstruction*)addr;
 }
 
-// The natural type of an RISCV instruction is uint32_t
-inline NativeInstruction* nativeInstruction_at(uint32_t *addr) {
-  return (NativeInstruction*)addr;
-}
+NativeCall* nativeCall_at(address addr);
+NativeCall* nativeCall_before(address return_address);
 
-inline NativeCall* nativeCall_at(address addr);
 // The NativeCall is an abstraction for accessing/manipulating native
 // call instructions (used to manipulate inline caches, primitive &
 // DSO calls, etc.).
-
-class NativeCall: public NativeInstruction {
+class NativeCall: private NativeInstruction {
  public:
-  enum RISCV_specific_constants {
-    instruction_size            =    4,
-    instruction_offset          =    0,
-    displacement_offset         =    0,
-    return_address_offset       =    4
+  enum {
+    instruction_size = 3 * Assembler::instruction_size,
```

Comment:
The issue is that the common code access the enum directly:
```
src/hotspot/share/code/nmethod.inline.hpp:    || (is_compiled_by_jvmci() && pc == (deopt_handler_begin() + NativeCall::instruction_size))
src/hotspot/share/code/nmethod.inline.hpp:    || (is_compiled_by_jvmci() && pc == (deopt_mh_handler_begin() + NativeCall::instruction_size))
src/hotspot/share/opto/output.cpp:  int pad_req   = NativeCall::instruction_size;

```

For c2 it just for an size estimate, so it's fine to give a larger value.

JVM CI is buggy (in general it's buggy all over), as we have no clue what size the unknown compiler using JVM CI generates here.
We can't use a constant for a C1/C2 call, so this code is just wrong.
Not sure what to do about, I have seen no errors for it.

Reply:
> The issue is that the common code access the enum directly:
> 
> ```
> src/hotspot/share/code/nmethod.inline.hpp:    || (is_compiled_by_jvmci() && pc == (deopt_handler_begin() + NativeCall::instruction_size))
> src/hotspot/share/code/nmethod.inline.hpp:    || (is_compiled_by_jvmci() && pc == (deopt_mh_handler_begin() + NativeCall::instruction_size))
> src/hotspot/share/opto/output.cpp:  int pad_req   = NativeCall::instruction_size;
> ```
> 
> For c2 it just for an size estimate, so it's fine to give a larger value.

Seems that hotspot shared code have some assumptions about the size of NativeCall. While it is OK for this c2 place, I am still a bit worried that we may have more uses of this const in hotspot shared code in future.

> JVM CI is buggy (in general it's buggy all over), as we have no clue what size the unknown compiler using JVM CI generates here. We can't use a constant for a C1/C2 call, so this code is just wrong. Not sure what to do about, I have seen no errors for it.

JVMCI is only *partially* supported on riscv for now and seems to be lack of love (See bug [1]).
I am not sure how this change will affect them.

[1] https://bugs.openjdk.org/browse/JDK-8290154

Reply:
Yea, we should probably add a `NativeCall::instruction_size()` but that limit us to have a fixed sized.
It would really be good if we could have variable size. E.g. look up size via pc desc.

Reply:
Sent out:
https://github.com/openjdk/jdk/pull/19556